src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java

```diff
@@ -233,6 +234,148 @@ private void deleteFile(final String pathName) {
+            command.append(str);
+            command.append(" ");
+        }
+        return command.toString().trim();
+    }
+
+    private void getDirContent(final ArrayList<String> commandArguments) {
+        boolean withAttrs = false;
+        Vector<ChannelSftp.LsEntry> fileAndDirectoryList;
+        try {
+            buildInfo.println(Messages.sftpExec_ls(getStringFromArrayList(commandArguments)));
+            String currentDir = sftp.pwd();
+            if (commandArguments.size() > 0) {
+                withAttrs = commandArguments.get(0).equals("-l");
+                if (withAttrs)
+                    commandArguments.remove(0);
+                if (commandArguments.size() > 0)
+                    currentDir = commandArguments.get(0);
+            }
+            fileAndDirectoryList = sftp.ls(currentDir);
+            if (fileAndDirectoryList.size() == 0)
+                buildInfo.println(Messages.sftpExec_lsEmpty(currentDir));
+            else
+                showDirContent(sftp.ls(currentDir), withAttrs);
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.sftpExec_failedCommand("ls"));
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+        }
+    }
+
+    private void showDirContent(final Vector<ChannelSftp.LsEntry> entry, final boolean withAttrs) {
+        for (ChannelSftp.LsEntry item : entry) {
+            final String getName;
+            if (!item.getFilename().equals(".") && !item.getFilename().equals("..")) {
+                if (withAttrs)
+                    getName = item.getLongname();
+                else
+                    getName = item.getFilename();
+                buildInfo.println(Messages.sftpExec_display(getName));
+            }
+        }
+        success();
+    }
+
+    private String parsePathName(String pathName, final boolean isDir) {
+        if (pathName.contains("*") || pathName.contains("?")) {
+            if (pathName.length() > 1 && pathName.contains("/"))
+                pathName = pathName.substring(0, pathName.lastIndexOf("/") + 1);
+            else
+                pathName = "./";
+        }
+        if (isDir && pathName.charAt(pathName.length() - 1) != '/')
+            pathName += "/";
+        return pathName;
+    }
+
+    private String parsePathName(String pathName) {
+        return parsePathName(pathName, true);
+    }
+
+    private boolean changeLocalBaseDirectory(final String newBaseDir) {
+        try {
+            buildInfo.printIfVerbose(Messages.sftpExec_lcd(newBaseDir));
+            sftp.lcd(newBaseDir);
+            success();
+            return true;
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+            return false;
+        }
+    }
+
+    private void getFiles(final ArrayList<String> commandArguments) {
+        String workspace = Util.replaceMacro("$WORKSPACE", buildInfo.getEnvVars());
```

Comment:
There are some cases with some types of jobs that this might not work correctly. The workspace should probably be passed in from the `perform` method somehow.

Reply:
Also, please use the interface types for arguments, e.g., `List<String>` instead of `ArrayList<String>`

Reply:
Can you explain in details what did you mean, please? As I understand, all environment variables are contained in `BPBuildInfo` instance. So `workspace` varibale is already passed in `BapSshClient` constructor from `perform` method by `BPBuildInfo` argument. Is it correct?

Reply:
I found that `buildInfo.getBaseDirectory` method is used to get local directory to put files listed in "Source files" field to remote host at the start of plugin work. Can I use this method inside BapSshClient to get workspace? Or the only one way is to rework BapSshClient constructor and pass workspace from `perform` method?

Reply:
Let me think about that. I will probably merge this in and rework later as necessary.